### PR TITLE
Ordering by index in select expression

### LINKED ISSUE #1414: `order by` supports index in select list
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
`order by` supports index, such as `select a from foo order by 1;`

**Describe the solution you'd like**
Add special process in https://github.com/apache/arrow-datafusion/blob/b9866e62800700dd322e755cfbbbff6b0df48fa2/datafusion/src/sql/planner.rs#L1011. Transfer `index` to `column` which is in select list. 

**Describe alternatives you've considered**
No

**Additional context**
No

### PR DESCRIPTION
**Pull request by hntd187**
# Which issue does this PR close?

Closes #1414 

 # Rationale for this change
You couldn't do this before, but now you can.

# What changes are included in this PR?

You can now order by the index of columns in a select statement. Sorry about the small formatting change, `cargo fmt` took over a bit there.

# Are there any user-facing changes?
New functionality for ordering by index

### CODE REVIEW

**Comment by xudong963** [comment]
Thanks @hntd187 , I'll review the ticket later today.

File: datafusion/src/sql/planner.rs [hunk 1]
```diff
@@ -2959,6 +2990,36 @@ mod tests {
         quick_test(sql, expected);
     }
 
+    #[test]
+    fn select_order_by_index() {
+        let sql = "SELECT id FROM person ORDER BY 1";
+        let expected = "Sort: #person.id ASC NULLS LAST\
+                        \n  Projection: #person.id\
```
Comment (liukun4515) [inline datafusion/src/sql/planner.rs:2997]: It is better to add more complex case:
```
select .... from table order by 1,3; // with multi column
```
```
select * from table order by 1,3
```

Comment (liukun4515) [reply datafusion/src/sql/planner.rs:2997]: You can add more tests after https://github.com/apache/arrow-datafusion/issues/1372 merged.

Comment (hntd187) [reply datafusion/src/sql/planner.rs:2997]: Sure I can add more

**Review by houqp — APPROVED** [review]
LGTM, agree with @liukun4515 's comment of adding more complex tests. Thanks @hntd187 .

File: datafusion/src/sql/planner.rs [hunk 2]
```diff
@@ -1026,14 +1025,44 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+            nulls_first,
+        } = e;
+
+        let expr = match &expr {
+            SQLExpr::Value(Value::Number(v, _)) => {
+                let field_index = v
+                    .parse::<usize>()
+                    .map_err(|err| DataFusionError::Plan(err.to_string()))?;
+
+                if field_index == 0 {
```
Comment (xudong963) [inline datafusion/src/sql/planner.rs:1040]: 👍

**Review by xudong963 — APPROVED** [review]
LGTM, thanks @hntd187

**Comment by alamb** [comment]
Thank you @hntd187  -- this feature (lack thereof) has annoyed me for a long time!

> You couldn't do this before, but now you can.

🤣

**Pull request merged**